File: 1_50/35_search_insert_position.java
class Solution {
    public int recurSearch(int[] nums, int low, int high, int target) {
        if (low > high) {
            // not found
            return low;
        }
        int mid = (low + high) / 2;
        
        if (nums[mid] > target) {
            return recurSearch(nums, low, mid-1, target);
        } else if (nums[mid] < target) {
            return recurSearch(nums, mid+1, high, target);
        } else {
            // found
            return mid;
        }
    }
    
    public int searchInsert(int[] nums, int target) {
        return recurSearch(nums, 0, nums.length-1, target);
    }
}

File: 1_50/35_search_insert_position.c
int searchInsert(int* nums, int numsSize, int target) {
    // use binary search
    int l = 0, r = numsSize - 1;
    int mid;
    
    while (l <= r) {
        mid = (l + r) / 2;
        if (nums[mid] > target) {
            r = mid - 1;
        } else if (nums[mid] < target) {
            l = mid + 1;
        } else {
            // found
            return mid;
        }
    }
    return l;
}
